### PR TITLE
Optimize type interfaces reading from metadata

### PR DESCRIPTION
**Pull request by DedSec256**
Continuation of https://github.com/dotnet/fsharp/pull/17364

As stated in https://github.com/dotnet/fsharp/pull/16168
image

In this PR, reading of interfaces and generic type parameters (also necessary for reading interfaces) is optimized similarly to the above-mentioned https://github.com/dotnet/fsharp/pull/17364 PR.

To verify the optimizations, all readings except for the readings of interfaces and generic parameters have been removed from `typeDefReader`.
The same benchmark data from https://github.com/dotnet/fsharp/pull/17364 was taken, and the following benchmark was written:

```fsharp
    [<Benchmark>]
    member _.ILReading() =
        for fileName in assemblies do
            let reader = OpenILModuleReader fileName readerOptions

            let ilModuleDef = reader.ILModuleDef

            for ilTypeDef in ilModuleDef.TypeDefs.AsArray() do
                ilTypeDef.Implements |> ignore

    [<IterationCleanup(Target = "ILReading")>]
    member _.ILReadingSetup() =
        // With caching, performance increases an order of magnitude when re-reading an ILModuleReader.
        // Clear it for benchmarking.
        ClearAllILModuleReaderCache()
```

> BenchmarkDotNet v0.13.10, Windows 11 (10.0.22631.3737/23H2/2023Update/SunValley3)
> 12th Gen Intel Core i9-12900H, 1 CPU, 20 logical and 14 physical cores
> .NET SDK 9.0.100-preview.5.24307.3
> [Host] : .NET 8.0.1 (8.0.123.58001), X64 RyuJIT AVX2 DEBUG
> Job-WKOUUV : .NET 8.0.1 (8.0.123.58001), X64 RyuJIT AVX2
> InvocationCount=1 UnrollFactor=1

| Method    | Mean     | Error   | StdDev   | Gen0      | Gen1      | Gen2      | Allocated |
|---------- |---------:|--------:|---------:|----------:|----------:|----------:|----------:|
| ILReading - OLD | 309.3 ms | 6.17 ms | 10.13 ms | 2000.0000 | 2000.0000 | 2000.0000 | 543.19 MB |
| ILReading - NEW | 288.2 ms | 5.62 ms | 9.08 ms | 2000.0000 | 2000.0000 | 2000.0000 | 507.47 MB |

As you can notice, the search phase for the interface range has become faster, as a smaller amount of data is being read. The full reading only occurs when converting the found rows:

image
*Profile for the sequential analysis of all files in [Giraffe](https://github.com/giraffe-fsharp/Giraffe/blob/master/src/Giraffe/Middleware.fs) - src/Giraffe*

### Plans

It is also planned to make the computation of type interfaces lazy in a separate PR, since it is independent optimization/API change.

### CODE REVIEW

File: src/Compiler/AbstractIL/ilread.fs [hunk 1]
```diff
@@ -878,10 +878,10 @@ let hfmCompare (TaggedIndex(t1: HasFieldMarshalTag, idx1)) (TaggedIndex(t2: HasF
...
     else compare t1.Tag t2.Tag
 
-let tomdCompare (TaggedIndex(t1: TypeOrMethodDefTag, idx1)) (TaggedIndex(t2: TypeOrMethodDefTag, idx2)) =
-    if idx1 < idx2 then -1
-    elif idx1 > idx2 then 1
-    else compare t1.Tag t2.Tag
+let inline tomdCompare (t1: TaggedIndex<TypeOrMethodDefTag>) (t2: TaggedIndex<TypeOrMethodDefTag>) =
+    if t1.index < t2.index then -1
+    elif t1.index > t2.index then 1
+    else compare t1.tag t2.tag
```
Comment (vzarytovskii) [inline src/Compiler/AbstractIL/ilread.fs:884]: Why not destructuring as before? This adds additional field/property access. Whereas before index was bound explicitly.

**Comment by edgarfgp** [comment]
@DedSec256 Thanks for improving this. Would you be able to fix the conflicts?. I think these performance improvements will make a difference in large projects what relies on interfaces and extension attributes.

**Comment by T-Gro** [comment]
@DedSec256 : We would be ready the merge this contribution now. Can you please resolve the new conflict Alex?

**Review by T-Gro — APPROVED** [review]

**Comment by DedSec256** [comment]
@T-Gro, There are some additional changes here related to reducing the copying of anonymous structures (and tuples creating) in private functions. Can I ask you to take yet another look?

File: src/Compiler/AbstractIL/ilread.fs (hunk 1 above)
Comment (DedSec256) [reply src/Compiler/AbstractIL/ilread.fs]: This gave some small perf improvement. But ok, removed from this PR

**Comment by DedSec256** [comment]
Looks flaky 🤔

image

File: src/Compiler/AbstractIL/ilread.fs [hunk 2]
```diff
@@ -2248,11 +2242,16 @@ and seekReadNestedTypeDefs (ctxt: ILMetadataReader) tidx =
...
+        id,
+        (fun idx ->
+            let mutable addr = ctxt.rowAddr TableNames.InterfaceImpl idx
+            let _tidx = seekReadUntaggedIdx TableNames.TypeDef ctxt mdv &addr
+            simpleIndexCompare tidx _tidx),
         isSorted ctxt TableNames.InterfaceImpl,
-        (fun x -> (seekReadTypeDefOrRef ctxt numTypars AsObject [] x.IntfIdx), (ctxt.customAttrsReader_InterfaceImpl, x.IntImplIdx))
+        (fun idx ->
+            let intfIdx = seekReadInterfaceIdx ctxt mdv idx
+            seekReadTypeDefOrRef ctxt numTypars AsObject [] intfIdx, (ctxt.customAttrsReader_InterfaceImpl, idx))
```
Comment (T-Gro) [inline src/Compiler/AbstractIL/ilread.fs:2254]: If this still reads the attributes applied on interface impls, then it's definitely good to get rid of  the anon from the signature!

**Review by T-Gro — APPROVED** [review]

File: src/Compiler/AbstractIL/ilread.fs (hunk 1 above)
Comment (vzarytovskii) [reply src/Compiler/AbstractIL/ilread.fs]: I was mostly curious why as it just looked as if it was adding field access vs binding directly

**Comment by psfinaki** [comment]
@DedSec256 before merging this - could you please update the benchmark numbers, just to make sure that they still hold after all this changed merged meanwhile?

**Comment by DedSec256** [comment]
@psfinaki, of course, I will try to do it soon.

**Comment by DedSec256** [comment]
| Method    | Mean     | Error   | StdDev   | Gen0      | Gen1      | Gen2      | Allocated |
|---------- |---------:|--------:|---------:|----------:|----------:|----------:|----------:|
| ILReading - Interfaces - Old (after nullnes merge) | 329.5 ms | 6.47 ms | 13.51 ms | 8000.0000 | 7000.0000 | 4000.0000 | 504.71 MB |
| ILReading - Interfaces - New | 312.6 ms | 6.23 ms | 11.08 ms | 8000.0000 | 7000.0000 | 4000.0000 | 501.87 MB |

**Comment by psfinaki** [comment]
Thanks @DedSec256 :)

**Review by psfinaki — APPROVED** [review]

**Review by edgarfgp — APPROVED** [review]

**Pull request merged**